Refactor FAB (Remove `RawMaterialButton`)

TahaTesser opened the pull request:
fixes https://github.com/flutter/flutter/issues/99086

### Description
This PR replaced `RawMaterialButton` with `ElevatedButton` (as it is obsolete and  soon to be deprecated)

There are golden changes due to highlight color not being translated to ElevatedButton (`ButtonStyleButton` sets highlight color to be [transparent](https://github.com/flutter/flutter/blob/6e2d31459d08d27e6076b087244b8bc7811687cf/packages/flutter/lib/src/material/button_style_button.dart#L368)).

### Highlight color golden changes preview (when FAB Is pressed)
|  M2 FAB (before/after) | M3 FAB  (before/after) |
| --------------- | --------------- |
| <img src="https://user-images.githubusercontent.com/48603081/157229979-be47089e-a8b2-4081-a433-76188ab25d80.png" /> | <img src="https://user-images.githubusercontent.com/48603081/157230028-65f06e46-8f1d-4ac1-9c86-553cff7df1f4.png" /> |



## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

TahaTesser committed:
Refactor FAB (Remove RawMaterialButton)

skia-gold commented:
Gold has detected about 4 new digest(s) on patchset 1.
View them at https://flutter-gold.skia.org/cl/github/99753

skia-gold commented:
Gold has detected about 6 new digest(s) on patchset 2.
View them at https://flutter-gold.skia.org/cl/github/99753

Piinks reviewed (commented):
Wow! Thank you for doing this. It looks like the golden file images show the color is slightly different, but I don't know if that is expected in the new design.

I am deferring to @HansMuller on the general design since he wrote the new Button/ButtonStyle universe. If this looks ok in keeping with the new paradigm I can give it a more thorough review.

I am wondering first if we are going to maintain all of the numerated properties on FAB, or deprecate and redirect them to be fashioned in a ButtonStyle. I think either approach has pros and cons, and this current one seems clean and minimally disruptive.
FAB already has its own FABTheme, so all of that churn would extend there as well. That being said I like this approach

HansMuller reviewed (commented):
I agree [with Kate](https://github.com/flutter/flutter/pull/99753#pullrequestreview-906492449) that just changing FAB's implementation, without adopting the ButtonStyle API here and in the FAB theme, is the prudent change to make.  I think that applies to ToggleButtons as well, per [Kate's feedback](https://github.com/flutter/flutter/pull/99493#pullrequestreview-906533911).

Supporting a ButtonStyle `style` parameter for FAB and ToggleButtons is a separate decision. If we choose to do that; it would be separate PR as well as a plan for migration, etc.

skia-gold commented:
Gold has detected about 6 new digest(s) on patchset 3.
View them at https://flutter-gold.skia.org/cl/github/99753

Piinks reviewed (commented):
Hey @TahaTesser what is the state of this PR? It looks like there are some failing tests.
I am concerned the subtle color change will cause a lot of internal tests to fail.

TahaTesser commented:
> Hey @TahaTesser what is the state of this PR? It looks like there are some failing tests.

Hey @Piinks 
We need to land https://github.com/flutter/flutter/issues/99975 in order to fix the tests.
These tests are were passing when I added Diagnosticable changes in this PR, which I reverted to be filed in a separate PR based on https://github.com/flutter/flutter/pull/99753#discussion_r824262494.

> I am concerned the subtle color change will cause a lot of internal tests to fail.

These golden changes are due to ButtonStyleButton setting `Colors.transparent` to highlight color when the button is pressed.
(Highlight color was provided in `RawMaterialButton`)

https://github.com/flutter/flutter/pull/99493 also has the same changes related highlight color.
cc @HansMuller could shed some light if this is okay or not.

Piinks commented:
Awesome! Thanks for the update, I was just checking that I was not neglecting this if it needed another review. :)

Piinks commented on packages/flutter/test/material/floating_action_button_theme_test.dart:
```diff
@@ -22,17 +22,25 @@ void main() {
-    expect(_getRawMaterialButton(tester).shape, const CircleBorder());
-    expect(_getRawMaterialButton(tester).splashColor, ThemeData().splashColor);
-    expect(_getRawMaterialButton(tester).constraints, const BoxConstraints.tightFor(width: 56.0, height: 56.0));
+    // // These defaults come directly from the [FloatingActionButton].
```
Nit: extra comment lines

Piinks commented on examples/api/test/material/floating_action_button/floating_action_button.0_test.dart:
```diff
@@ -16,11 +16,11 @@ void main() {
-    }
-    expect(getRawMaterialButtonWidget().fillColor, Colors.green);
-    expect(getRawMaterialButtonWidget().shape, const CircleBorder());
+    final Material material = tester.widget(find.descendant(
```
There may be tests like this in google that need to be migrated, like the ToggleButton change. I am running internal tests now. :)

Piinks commented on packages/flutter/test/material/floating_action_button_theme_test.dart:
```diff
@@ -22,17 +22,25 @@ void main() {
 
+    final ElevatedButton elevatedButton = tester.widget(find.byType(ElevatedButton));
+    final BoxConstraints buttonConstraints = BoxConstraints(
+      minWidth: elevatedButton.style!.minimumSize!.resolve(enabled)!.width,
+      minHeight: elevatedButton.style!.minimumSize!.resolve(enabled)!.height,
+      maxWidth: elevatedButton.style!.maximumSize!.resolve(enabled)!.width,
+      maxHeight: elevatedButton.style!.maximumSize!.resolve(enabled)!.height,
```
Here and elsewhere, can you store some of these, like `elevatedButton.style`,  in local variables for readability? You would not have to have so many `!`s either. :)

Piinks commented on packages/flutter/test/material/floating_action_button_theme_test.dart:
```diff
@@ -22,17 +22,25 @@ void main() {
       ),
     ));
 
+    final ElevatedButton elevatedButton = tester.widget(find.byType(ElevatedButton));
```
This could even simplify things a bit:
```suggestion
    final ButtonStyle style = tester.widget(find.byType(ElevatedButton)).style!;
```

Piinks reviewed (commented):
This looks great! Just some style nits. I am checking with customer tests.

Piinks commented:
There are indeed some customers to migrate, working on that now. I'll update to land once complete. :)

Piinks commented:
Update: still working with customers. There are some changes here that cause the color of the button to fail contrast guidelines when pressed. That may be related to the highlight color change you mentioned above, still tracking it down.

Piinks commented:
Not forgotten. :) Just back from time off, following up on customer migrations. I'll report back tomorrow.

Piinks commented:
@TahaTesser can you update this branch with the latest from master? I am having trouble using our tools to sync this with the internal copy for testing. Thanks!

Piinks commented:
Thanks for updating! This is still going to take a minute to fix all of the affected customers, we're making progress though!

TahaTesser commented:
@Piinks 
Fixed conflicts and rebased this too.

Piinks commented:
Thanks for the update! This is now at the front of my customer migration list.

Piinks commented:
Welcome back from vacation again. :) This is under active migration with customers.

TahaTesser commented:
@Piinks 
Thank you so much.  Looking forward to merging this :)

Piinks commented on packages/flutter/lib/src/material/floating_action_button.dart:
```diff
@@ -602,26 +605,39 @@ class FloatingActionButton extends StatelessWidget {
+        padding: MaterialStateProperty.all<EdgeInsetsGeometry?>(EdgeInsets.zero),
+        minimumSize: MaterialStateProperty.all<Size?>(Size(sizeConstraints.minWidth, sizeConstraints.minHeight)),
+        maximumSize: MaterialStateProperty.all<Size?>(Size(sizeConstraints.maxWidth, sizeConstraints.maxHeight)),
+        shape: MaterialStateProperty.all<OutlinedBorder?>(shape as OutlinedBorder),
```
```suggestion
        shape: MaterialStateProperty.all<OutlinedBorder?>(shape as OutlinedBorder),
```

This is causing issues for customers that have 

```
FloatingActionButton(
  shape: Border(),
  ...
)
```

Is there a way that this can still be supported? Or is it a breaking change?

TahaTesser replied:
It can still be supported by wrapping `ElevatedButton` with `DecoratedBox` BUT this can be done by the customer too, `floatingActionButton:` takes any widget so FAB can be wrapped in a `DecoratedBox`.

The reason when  `Border()`  provide to the `shape` parameter currently works is because `material.dart` checks if "shape" is a `ShapeBorder` otherwise based on material type, directly draw the border with custom paint hence there is no issue with this in the current design.


https://github.com/flutter/flutter/blob/75ee440329f555cac3d636a24bdc5c16c8ef60ef/packages/flutter/lib/src/material/material.dart#L908-L916

but there is no existing test where the border is provided to shape in the test class and looks a little bit off to me when the border is provided to shape parameter. 

If you still want this, I can wrap  `ElevatedButton `with `DecoratedBox` along with a couple of tests to make sure this behavior works as expected.

Piinks replied:
> If you still want this, I can wrap ElevatedButton with DecoratedBox along with a couple of tests to make sure this behavior works as expected.

There are a surprising number of cases with Border, if we can still support it, it would be definitely be easier. Do you want to try it and see what it would look like?

TahaTesser replied:
I just added the code. 
`decoration: ShapeDecoration(shape: shape),` is responsible for rendering the border and we're defaulting to a rectangular shape to match current behavior when a border is provided.

```dart
          shape: MaterialStateProperty.all<OutlinedBorder?>(shape is! OutlinedBorder
            ? const RoundedRectangleBorder()
            : shape as OutlinedBorder?,
          ),
```

![Screenshot 2022-08-22 at 11 25 48](https://user-images.githubusercontent.com/48603081/185952579-5574a5ba-d5e6-4b2d-89ea-97ac53f78226.png)

TahaTesser committed:
Add decorated box

Piinks commented:
Still catching up from having covid, thanks for your patience. For this, there is one more customer test file that I need to migrate, and then this should be ready.

TahaTesser commented:
@Piinks 
Friendly ping.

TahaTesser commented:
Closing this PR for now. Will come to this with a clean PR in the future.
